README.md
# Turtle Autocompleter package
[![dependencies Status](https://david-dm.org/thomsonreuters/AtomOntologyCompleter/status.svg?style=flat-square)](https://david-dm.org/thomsonreuters/AtomOntologyCompleter)
[![Build Status](https://travis-ci.org/thomsonreuters/AtomOntologyCompleter.svg?branch=master&style=flat-square)](https://travis-ci.org/thomsonreuters/AtomOntologyCompleter)
[![Installs!](https://img.shields.io/apm/dm/atom-clock.svg?style=flat-square)](https://atom.io/packages/turtle-completer)
[![Version!](https://img.shields.io/apm/v/atom-clock.svg?style=flat-square)](https://atom.io/packages/turtle-completer)
[![License](https://img.shields.io/apm/l/atom-clock.svg?style=flat-square)](https://github.com/thomsonreuters/AtomOntologyCompleter/blob/master/LICENSE)
[![Gitter chat](https://badges.gitter.im/thomsonreuters/AtomOntologyCompleter.svg?style=flat-square)](https://gitter.im/thomsonreuters/AtomOntologyCompleter)

This is an autocompleter for the [Atom](http://atom.io) editor, compatible with [AutoComplete+](https://atom.io/packages/autocomplete-plus) that knows common ontologies and will suggest predicates based on your typing. Both the [Turtle](https://www.w3.org/TR/turtle/) and [TriG](https://www.w3.org/TR/trig/) serializations are supported.

![Short video of the completer in action](/docs/completer-sample-video.gif?raw=true)

Current ontologies are:
* [Activity Vocabulary](http://dublincore.org/documents/dcmi-terms/)
* [Metadata for Tabular Data](https://www.w3.org/TR/tabular-metadata/)
* [Data Catalog Vocabulary](https://www.w3.org/TR/vocab-dcat/)
* [Data Cubes](https://www.w3.org/TR/vocab-data-cube/)
* [Data Quality Vocabulary](http://www.w3.org/TR/vocab-dqv)
* [Data Usage Vocabulary](https://www.w3.org/TR/vocab-duv/#)
* [Dublin Core & Terms](http://dublincore.org/documents/dcmi-terms/)
* [Friend of a Friend](http://xmlns.com/foaf/spec/)
* [ODRL](https://www.w3.org/TR/odrl-vocab/)
* [Ontology for Media Resources](https://www.w3.org/TR/2012/REC-mediaont-10-20120209/)
* [OWL](https://www.w3.org/TR/owl2-overview/)
* [R2RML](http://www.w3.org/TR/2012/REC-r2rml-20120927/)
* [RDF](https://www.w3.org/TR/2004/REC-rdf-mt-20040210/)
* [RDFS](https://www.w3.org/TR/rdf-schema/)
* [Relationship](http://www.perceive.net/schemas/20031015/relationship/)
* [SHapes Constraint Language](https://www.w3.org/TR/shacl/)
* [SKOS](https://www.w3.org/2009/08/skos-reference/skos.html)
* [W3C Provenance](https://www.w3.org/TR/2013/REC-prov-o-20130430/)
* [Web Annotation Vocabulary](http://www.w3.org/TR/annotation-vocab/)

If you'd like others added either raise an issue on the atom package, comment on gitter, tweet @nonodename or do a pull request with the conversion. Contributions are greatly appreciated.  Note that the ontology is matched based on common prefixes.

Known ontologies are referenced from the file `ontologies.json` in `\lib`. Each ontology has a JSON file that contains the completion info.

To generate a new file, run `convert.py` over a [ttl](https://en.wikipedia.org/wiki/Turtle_(syntax)) version of the ontology to generate the JSON. convert.py supports a few command line options, run without any parameters to see them.

Note that quality of anchor links into the specs varies on whether the naming convention for the anchors is sufficiently predictable. Feel free to fix any that are wrong in the relevant JSON file and do a pull request
to get added.

## Testing ontologies
To test a new ontology without the palaver of pull requests:
* Create a copy of the file ontologies.json in ``\lib``
* Convert your ontology to the JSON format required for the completer
* Update the ontologies.json file, setting the key to the desired prefix and
ontology attribute to the filename & path of the converted ontology.
* The spec, prefix and docprefix attributes aren't part of the code path so can
be ignored or completed as you wish
* Note that if you re-use a predefined prefix you'll override the built in
ontology

For example:
```
"tr":{
  "title":"My First ontology",
  "spec":"URL of the spec",
  "prefix":"tr",
  "docprefix":"URL prefix for documentation",
  "ontology":"/users/dan/ontologies/tr.json"
}
```


CHANGELOG.md
## 1.0.0
* Now supports the TriG serialization in addition to Turtle
## 0.6.0
* Fixed bug that suggestions wouldn't be provided mid-line
* Fixed bug in converter that wouldn't allow predicates with a number in. Thanks again to Stefan for reporting.
* More unit tests
* Added the relationship ontology

## 0.5.1
* Code and documentation clean up

## 0.5.0
* Added capability to load local ontologies
* Fixed convert.py to work on Python 2 or 3 (thanks again to Stefan for reporting that)

## 0.4.2
* Fixed rdf.json. Thanks to Stefan Münnich for the issue

## 0.4.1
* Updated README, patched for ontologies missed on checkin

## 0.4.0
* Added unit tests
* Now builds on travis.ci so we know the tests work
* Added OWL, Ontology for Media Resources, Data Usage Vocabulary, Web Annotation Vocabulary
* Data Cubes, RDF, R2RML, ODRL

## 0.3.0
* The growth of ontologies is slowing load time. Lazy loading, engage!
* Sometimes the completions wouldn't be accepted when you press enter. After rewriting in CoffeeScript I discovered that this is because you need to pass a copy of your suggestions to the provider API otherwise it seems to get confused

## 0.2.0
* Added Activity Streams, Tabular Metadata, DCAT, Data Quality Vocabulary

## 0.1.1
* License fixes, improved readme

## 0.1.0
* Initial release


lib/turtle-provider.js
/* Copyright 2018 Thomson Reuters
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
const ontologies = require("./ontologies.json");

module.exports =
class TurtleProvider {
  constructor() {
    this.selector = '.source.turtle'
    this.disableForSelector = '.source.turtle .comment'
    this.inclusionPriority = 1
    this.suggestionPriority = 1
    //this.excludeLowerPriority = true
    this.filterSuggestions = false
    this.ontologyCompletions = [];
  //  this.showIcon = ['Symbol', 'Subsequence'].includes(atom.config.get('autocomplete-plus.defaultProvider'))
  }

  loadOntologies(){
    var completions = [];
    Object.keys(ontologies).forEach(function(key, index){
      var ontologyEntry = {};
      ontologyEntry.description = ontologies[key].title;
      ontologyEntry.descriptionMoreURL = ontologies[key].spec;
      ontologyEntry.text = key;
      completions.push(ontologyEntry);
    },ontologies);
    this.ontologyCompletions = completions;
  }


  loadOntology(ontologyPrefix){
    if(ontologyPrefix in ontologies){
      var completions = require(ontologies[ontologyPrefix].ontology);
      ontologies[ontologyPrefix].completions = completions;
    }
  }

  getSuggestions({editor,bufferPosition,scopeDescriptor, prefix}) {
    if (!(prefix != null ? prefix.length : undefined)) { return []}
    var currentScope = scopeDescriptor.getScopesArray();
    var candidateArray = null;
    if(currentScope.length < 2){ // we're not sure of scope yet, so choose from ontologies.
      candidateArray = this.ontologyCompletions;
    } else if (currentScope.includes("entity.name.tag.prefixed-uri.turtle")) {
      var namespace = this.determineNamespace(editor.lineTextForBufferRow(bufferPosition.row));
      if(namespace && namespace in ontologies)
        if(ontologies[namespace].completions == null)
          this.loadOntology(namespace);
        candidateArray = ontologies[namespace].completions;
    }
    return this.getMatches(candidateArray,prefix);

  }

  getMatches(candidateArray, prefix){
    if(candidateArray == null) {return []}
    var matches = [];
    candidateArray.forEach(function(element){
      if(firstCharsEqual(element.text,prefix))
        matches.push(element);
    });
    return matches;
  }
  determineNamespace(line, position){
    var regex = /([\w]+):[^:]+$/;
    var match = regex.exec(line);
    return match[1];
  }

/*  onDidInsertSuggestion({editor}) {
    return atom.commands.dispatch(atom.views.getView(editor), 'snippets:expand')
  }*/
}

const firstCharsEqual = (str1, str2) => str1[0].toLowerCase() === str2[0].toLowerCase()


lib/ontologies/convert.py
# Copyright 2018 Thomson Reuters

# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

# Convert a turtle file to a JSON object suitable for the autocompleter
# For example:
# cat foaf.ttl | python convert.py "http://xmlns.com/foaf/spec/#" foaf > foaf.json

# Prefixes for documentation are found in ontologies.json
# Push turtle via cat
# Only tested on Mac OS with Python 2.7

import sys
import re
import json

def compile_regex(pattern):
    if hasattr(pattern, 'decode'):
        # Python 2, decode to unicode first
        return re.compile(pattern.decode('ascii'))
    else:
        return re.compile(pattern)

if len(sys.argv) < 3:
    sys.exit('Usage %s <documentation_url_base> <desired_prefix_to_extract>' % sys.argv[0])
urlPrefix=sys.argv[1]
desiredPrefix=sys.argv[2]
predicates = []
predicateRegex = compile_regex(r'^([^:\s]*):([\w]+)')
typeRegex = compile_regex(r'a (rdf|owl|rdfs|dcam):([\w]+)')
commentRegex = compile_regex(r'\s+(rdfs:comment|skos:definition|rdfs:label) \"([^\"]+)\"(@\w\w)?')
currentPredicate = None
for line in sys.stdin:
    predicateName = predicateRegex.match(line)
    if predicateName is not None and predicateName.group(1) == desiredPrefix:
        if currentPredicate is not None:
            currentPredicate['descriptionMoreURL']=urlPrefix+currentPredicate['text']
            predicates.append(currentPredicate)
        currentPredicate = {}
        currentPredicate['text'] = predicateName.group(2)
        # continue - no continue here since some specs the type is on the same line
    if currentPredicate is not None:
        predicateType = typeRegex.search(line)
        if predicateType is not None and 'type' not in currentPredicate:
            if predicateType.group(2) == 'Class' \
                or predicateType.group(2) == 'Datatype' \
                or predicateType.group(2) == 'VocabularyEncodingScheme':
                currentPredicate['type']='class'
            elif 'Property' in predicateType.group(2): # bit of a hack
                currentPredicate['type']='property'
            elif predicateType.group(2) == 'Resource':
                currentPredicate['type']='snippet'
            else:
                currentPredicate['type']='Something weird: '+predicateType.group(2)
            continue
        predicateComment = commentRegex.match(line)
        if predicateComment is not None and (predicateComment.group(3) is None or predicateComment.group(3) == '@en'):
            if 'description' not in currentPredicate:
                currentPredicate['description']=predicateComment.group(2)
            else:
                if predicateComment.group(1) != 'rdfs:label':   # permit rdfs:label as a fallback
                    currentPredicate['description']=predicateComment.group(2)

json.dump(predicates,sys.stdout,indent=3,sort_keys=True)


lib/autocomplete-turtle.js
/* Copyright 2018 Thomson Reuters
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
module.exports = {
  provider: null,

  activate() {},

  deactivate() {
    this.provider = null
  },

  provide() {
    if (this.provider == null) {
      const TurtleProvider = require('./turtle-provider')
      this.provider = new TurtleProvider()
      if(this.provider != null){
        this.provider.loadOntologies();
      }
    }

    return this.provider
  },

}


lib/ontologies/generate.sh
# Run this script to generate the json configurationfor autocompletion
# The ontologies themselves are available from source, not stored in the project as duplicative

cat as.ttl | python convert.py "https://www.w3.org/TR/activitystreams-vocabulary/#dfn-" as > as.json
cat csvw.ttl | python convert.py "https://www.w3.org/TR/tabular-metadata/#" csvw > csvw.json
cat dc11.ttl | python convert.py "http://dublincore.org/documents/dcmi-terms/#terms-" dc11 > dc11.json
cat dcat.ttl | python convert.py "https://www.w3.org/TR/vocab-dcat/#" dcat > dcat.json
cat dcterms.ttl | python convert.py "http://dublincore.org/documents/dcmi-terms/#elements-" dcterms > dcterms.json
cat dqv.ttl | python convert.py "https://www.w3.org/TR/vocab-dqv/#dqv:" dqv > dqv.json
cat duv.ttl | python convert.py "https://www.w3.org/TR/vocab-duv/#" duv > duv.json
cat foaf.ttl | python convert.py "http://xmlns.com/foaf/spec/#" foaf > foaf.json
cat ma.ttl | python convert.py "https://www.w3.org/TR/2012/REC-mediaont-10-20120209/#" ma > ma.json
cat oa.ttl | python convert.py "https://www.w3.org/TR/annotation-vocab/#" oa > oa.json
cat odrl.ttl | python convert.py "https://www.w3.org/TR/odrl-vocab/#term-" odrl > odrl.json
cat owl.ttl | python convert.py "https://www.w3.org/TR/2012/REC-owl2-syntax-20121211/#a_" owl > owl.json
cat prov.ttl | python convert.py "https://www.w3.org/TR/2013/REC-prov-o-20130430/#" "" > prov.json
cat qb.ttl | python convert.py "https://www.w3.org/TR/vocab-data-cube/#dfn-qb-" qb > qb.json
cat rdf.ttl | python convert.py "https://www.w3.org/TR/2004/REC-rdf-mt-20040210/#" rdf > rdf.json
cat rdfs.ttl | python convert.py "https://www.w3.org/TR/rdf-schema/#ch_" rdfs > rdfs.json
cat ttl/rel.ttl | python convert.py "http://www.perceive.net/schemas/20031015/relationship/" rel > rel.json
cat rr.ttl | python convert.py "https://www.w3.org/ns/r2rml#" rr > rr.json
cat shacl.ttl | python convert.py "https://www.w3.org/TR/shacl/#" sh > shacl.json
cat skos.ttl | python convert.py "https://www.w3.org/2009/08/skos-reference/skos.html#" skos > skos.json
